### PR TITLE
Add gateway dependencies and setup CI

### PR DESCRIPTION
**Pull request by harupy**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Add gateway dependencies and setup CI.

## How is this patch tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: .github/workflows/gateway.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+name: Gateway
```
Comment (harupy) [inline .github/workflows/gateway.yml:1]: Using a separate workflow for faster iteration. We can remove this later.

File: .github/workflows/gateway.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  gateway:
+    runs-on: ubuntu-latest
+    timeout-minutes: 30
+    steps:
+      - uses: actions/checkout@v3
+      - uses: ./.github/actions/setup-python
+      - name: Install dependencies
+        run: |
+          pip install -e .[gateway]
+          pip install pytest pytest-timeout
```
Comment (harupy) [inline .github/workflows/gateway.yml:28]: Minimum dependencies to develop faster.

File: .github/workflows/gateway.yml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
+
+defaults:
+  run:
+    shell: bash --noprofile --norc -exo pipefail {0}
+
+jobs:
+  gateway:
+    runs-on: ubuntu-latest
```
Comment (harupy) [inline .github/workflows/gateway.yml:20]: We could add windows later.

File: tests/gateway/test_gateway_handlers.py [hunk 4]
```diff
@@ -0,0 +1,2 @@
+def test_nothing():
+    pass
```
Comment (harupy) [inline tests/gateway/test_gateway_handlers.py:2]: A dummy test. pytest exits with non-zero when there are no tests to run.

File: setup.py [hunk 5]
```diff
@@ -160,6 +160,11 @@ def run(self):
             "google-cloud-storage>=1.30.0",
             "boto3>1",
         ],
+        "gateway": [
+            "fastapi",
+            "pydantic",
+            "uvicorn",
```
Comment (harupy) [inline setup.py:166]: We can pin or add new dependencies later if necessary.

Comment (dbczumar) [reply setup.py:166]: Are there some reasonable major version lower and upper bounds we can set?

**Review by dbczumar — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/8619/files#r1219027649 is addressed. Thanks @harupy !

File: setup.py (hunk 5 above)
Comment (harupy) [reply setup.py:166]: Can we just use the same rule as the core dependencies (using `< current_latest_version + 1`)?

Comment (harupy) [reply setup.py:166]: we can create `gateway-requirements.yml` if necessary.

Comment (harupy) [reply setup.py:166]: The current latest version of each package:

- https://pypi.org/project/fastapi/0.96.0/
- https://pypi.org/project/uvicorn/0.22.0/
- https://pypi.org/project/pydantic/2.0b2/

Comment (harupy) [reply setup.py:166]: Do we really need to pin these?

Comment (dbczumar) [reply setup.py:166]: Yes, it's our policy to do pinning on major versions now + compatibility tests. I think we should do the same for gateway

Comment (dbczumar) [reply setup.py:166]: For pydantic, shall we add a lower version bound (`>=1`)?

Comment (harupy) [reply setup.py:166]: Sounds good. I'll add `gateway-requirements.yml`.

File: requirements/gateway-requirements.yaml [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  minimum: "1.0"
+  max_major_version: 2
+
+fastapi:
+  pip_release: fastapi
+  max_major_version: 0
+
+uvicorn:
+  pip_release: uvicorn
+  max_major_version: 0
```
Comment (harupy) [inline requirements/gateway-requirements.yaml:15]: @dbczumar Added this file.

**Review by BenWilson2 — APPROVED** [review]
LGTM! Thanks @harupy :)

**Pull request merged**